File: Data Structures & Algorithms/longest-common-subsequence/submission-1.java
class Solution {
    public int longestCommonSubsequence(String s, String t) {
        if(s.length() < t.length()){
            return longestCommonSubsequence(t,s);
        }

        int m = s.length(), n = t.length(), dp[] = new int[n+1];
        dp[n] = 0;

        for(int i = m - 1; i >= 0; --i){
            int prev = 0;
            for(int j = n - 1; j >= 0; --j){
                int temp = dp[j];
                if(s.charAt(i) == t.charAt(j)){
                    dp[j] = 1 + prev;
                }
                else{
                    dp[j] = Math.max(dp[j], dp[j+1]);
                }
                prev = temp;
            }
        }
        return dp[0];
    }
}


/*

i : m-1 -> 0
j : n-1 -> 0
case 1: s[i] == t[j] :-
    T[i][j] = 1 + T[i+1][j+1]
case 2: s[i] != t[j] :-
    T[i][j] = Math.max(T[i+1][j], T[i][j+1])

*/

File: Data Structures & Algorithms/longest-common-subsequence/submission-1.py
class Solution:
    def longestCommonSubsequence(self, s: str, t: str) -> int:
        """
        if chars are equal
        fn(i,j) = 1 + fn(i-1, j-1)
        else
        fn(i,j) = max(fn(i-1,j), fn(i,j-1))

        in terms of arrays 
        equal case
        dp[j] = 1 + dp[j-1]
        else case
        dp[j] = max(dp[j], dp[j-1])
        """

        m, n = len(s), len(t)
        prev = [0] * (n+1)

        for i in range(1, m+1):
            cur = [0] * (n+1)
            for j in range(1, n+1):
                if s[i-1] == t[j-1]:
                    cur[j] = 1 + prev[j-1]
                else:
                    cur[j] = max(cur[j-1], prev[j])
            prev = cur
        return prev[n]

        